### PR TITLE
Solving N+1 Problem and Adding Capybara Integration Tests

### PR DESCRIPTION
**Pull request by christianonoh**
### Description
This pull request addresses two main aspects of the project requirements:

1. **Solving N+1 Problem**: The N+1 problem was causing performance issues when fetching all posts and their comments for a user. To solve this problem, we have refactored the code to use eager loading for fetching posts and comments, significantly improving the application's performance. We have analyzed the logs following the guidelines in the article provided and ensured that the N+1 problem is resolved.

2. **Integration Tests using Capybara**: We have added integration tests using Capybara to ensure that the user interface behaves as expected for each view in the project. The tests are grouped logically to cover different functionalities and interactions with the application. Each test verifies specific elements and interactions on the respective pages, ensuring that the app's UI functions correctly.

#### N+1 Problem Fix:
- Refactored the code to use eager loading for fetching posts and their comments for a user.
- Analyzed logs before and after the fix to demonstrate the performance improvement.

#### Capybara Integration Tests:
1. User Index Page:
   - Test to verify the presence of usernames of all other users.
   - Test to verify the display of profile pictures for each user.
   - Test to check the correct number of posts written by each user.
   - Test to ensure clicking on a user redirects to their show page.

2. User Show Page:
   - Test to verify the presence of the user's profile picture.
   - Test to check the display of the user's username.
   - Test to ensure the correct number of posts written by the user.
   - Test to verify the display of the user's bio.
   - Test to ensure the display of the user's first 3 posts.
   - Test to check the functionality of the button to view all posts.
   - Test to ensure clicking on a user's post redirects to the post's show page.
   - Test to verify that clicking on the "view all posts" button redirects to the user's post index page.

3. User Post Index Page:
   - Test to verify the presence of the user's profile picture.
   - Test to check the display of the user's username.
   - Test to ensure the correct number of posts written by the user.
   - Test to verify the display of a post's title.
   - Test to check the display of some of the post's body.
   - Test to ensure the display of the first comments on a post.
   - Test to verify the number of comments a post has.
   - Test to verify the number of likes a post has.
   - Test to check the presence of a pagination section if there are more posts than fit on the view.
   - Test to ensure clicking on a post redirects to the post's show page.

4. Post Show Page:
   - Test to verify the presence of the post's title.
   - Test to check the display of the post's author.
   - Test to ensure the correct number of comments on the post.
   - Test to verify the number of likes on the post.
   - Test to check the display of the post body.
   - Test to ensure the display of the usernames of each commenter.
   - Test to verify the display of the comments left by each commenter.

### Screenshots
Please find attached the screenshots of the app's logs before
Screenshot 2023-07-25 at 6 30 28 AM
 and after the N+1 problem fix.
Screenshot 2023-07-25 at 6 38 23 AM

### Testing
All integration tests have been created using Capybara, and RSpec has been used to run the test suite. The tests cover the specified functionalities and interactions as per the project requirements.

**Please review the changes and test results and give us your feeback**

Thank you for your time and consideration!

Best regards,
@christianonoh & @ellyboi

### CODE REVIEW

**Review by omar-muhamad — COMMENTED** [review]
# Invalid ❌

Hi @christianonoh and @Ellyboi 

#### While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

#### Your Code Review Request is marked as invalid for this reason:

- [ ] Some linters checks are not passed. Please make sure that all checks are passing and green.


Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **Does not count** into the code reviews limit._

File: spec/integration/user_index_spec.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  end
+
+  it 'displays the username of all other users' do
+    expect(page).to have_content('Kotonko')
+    expect(page).to have_content('Usom')
+  end
+
+  it 'should have profile picture content of all other users' do
+    expect(page).to have_css("img[src*='/images/default.jpg']")
+    expect(page).to have_css("img[src*='/images/default.jpg']")
```
Comment (TadesseDev) [inline spec/integration/user_index_spec.rb]: - [x] Dear Team, here we are trying to test if we can see the profile picture of `each` user. This means we need to test the profile picture with a different `src` value. 
- [x] Besides that, I can see you are trying to find a picture with `src*='/images/default.jpg` while in fact, the value of a `photo` attribute for a user is ` photo: 'https://i.picsum.photos/id/216/200/300.jpg?hmac'`. We need to test whether the value is the same as what the user has in the database.

File: spec/integration/user_show_spec.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+      @user2 = User.create(name: 'Burger', photo: 'https://i.picsum.photos/id/216/200/300.jpg?hmac', bio: 'Burger bio')
+      @post1 = Post.create(title: 'Post 1', text: 'Post content goes here', author_id: @user1.id)
+      @post2 = Post.create(title: 'Post 2', text: 'Post content goes here', author_id: @user1.id)
+      @post3 = Post.create(title: 'Post 3', text: 'Post content goes here', author_id: @user2.id)
+      visit user_path(@user1)
+    end
+
+    it 'should have profile picture content of all other users' do
+      expect(page).to have_css("img[src*='/images/default.jpg']")
+    end
```
Comment (TadesseDev) [inline spec/integration/user_show_spec.rb]: - [x] Please follow up on my recommendation on how to test a profile picture of a user in a previous comment. Please make sure to test the `photo: value` in the `src: value`.

File: spec/integration/user_show_spec.rb [hunk 3]
```diff
@@ -0,0 +1,31 @@
+require 'rails_helper'
+
+RSpec.describe 'Specific User', type: :feature do
+  describe 'Show Page' do
+    before(:each) do
+      @user1 = User.create(name: 'Usom', photo: 'https://source.unsplash.com/user/c_v_r/100x100', bio: 'Usom bio')
```
Comment (TadesseDev) [inline spec/integration/user_show_spec.rb]: #### Dear Team, for a user_show route some tests are missing please consider adding them. 
- [x] Tests that assert you can see the number of posts the user has written.
- [x] Tests that assert you can see a button that lets me view all of a user's posts.
- [x] Tests that redirect to the post's show page when you click on a user's post.
- [x] Tests that redirects to the user's post's index page when you click on a **see all posts**.

File: spec/integration/post_index_spec.rb [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+      @like1 = Like.create(author_id: @user1.id, post_id: @post1.id)
+      @like2 = Like.create(author_id: @user1.id, post_id: @post1.id)
+      @like3 = Like.create(author_id: @user2.id, post_id: @post1.id)
+
+      visit user_posts_path(@user1)
+    end
+
+    it 'Should have profile picture content of all other users' do
+      expect(page).to have_css("img[src*='/images/default.jpg']")
+    end
```
Comment (TadesseDev) [inline spec/integration/post_index_spec.rb]: - [x] Please follow up on my recommendation on how to test a profile picture of a user in a previous comment. Please make sure to test the `photo: value` in the `src: value`.

File: spec/integration/post_index_spec.rb [hunk 5]
```diff
@@ -0,0 +1,58 @@
+require 'rails_helper'
+
+RSpec.describe 'User posts', type: :feature do
+  describe 'Show Page' do
+    before(:each) do
+      @user1 = User.create(name: 'Usom', photo: 'https://source.unsplash.com/user/c_v_r/100x100', bio: 'Usom bio')
+      @user2 = User.create(name: 'Burger', photo: 'https://i.picsum.photos/id/216/200/300.jpg?hmac', bio: 'Burger bio')
```
Comment (TadesseDev) [inline spec/integration/post_index_spec.rb]: #### Dear Team, for a post_index route some tests are missing please consider adding them. 
- [x] Tests that asserts you can see the number of posts the user has written.
- [x] Tests that asserts you can see a post's title.

**Review by TadesseDev — CHANGES_REQUESTED** [review]
### Hi Team

### I'm Tadesse Alemayehu your reviewer for this PR. :hand: 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!
212094426-b41dd457-c9db-4ab3-bbe9-176dd8dd08b7




## Highlights
 - [x]  Most of the functionality has been tested. ✅
 - [x]  Great job on solving the N+1 problem on the post controller (consider applying this in all the other appropriate places) ✅
 
## Required Changes ♻️
- [x] Please mention both the collaborator in the `Authors` section of the readme. 
_Check the comments under the review._

## Optional suggestions
- N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by codepantha — APPROVED** [review]
# STATUS: APPROVED 🟢 ✔️ 

## Hi team 👋, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



## Highlights 🎯 
✔️ Descriptive PR
✔️ Solved n+1 problem 📌 
✔️ All tests are green ✔️ 

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by christianonoh** [comment]
Thank you! @codepantha

**Pull request merged**